pkg/component/runtime/runtime_comm.go

```diff
@@ -225,85 +240,93 @@ func (c *runtimeComm) checkin(server proto.ElasticAgent_CheckinV2Server, init *p
 		c.checkinLock.Unlock()
 	}()
 
-	initExp := make(chan *proto.CheckinExpected)
-	recvDone := make(chan bool)
-	sendDone := make(chan bool)
-	go func() {
-		defer func() {
-			close(sendDone)
-		}()
-
-		// initial startup waits for the first expected message from the dedicated initExp channel
-		select {
-		case <-checkinDone:
-			return
-		case <-recvDone:
-			return
-		case expected := <-initExp:
-			err := sendExpectedChunked(server, expected, c.chunkingAllowed, c.maxMessageSize)
-			if err != nil {
-				if reportableErr(err) {
-					c.logger.Debugf("check-in stream failed to send initial expected state: %s", err)
-				}
-				return
-			}
-		}
-
-		for {
-			var expected *proto.CheckinExpected
-			select {
-			case <-checkinDone:
-				return
-			case <-recvDone:
-				return
-			case expected = <-c.checkinExpected:
-			}
-
-			err := sendExpectedChunked(server, expected, c.chunkingAllowed, c.maxMessageSize)
-			if err != nil {
-				if reportableErr(err) {
-					c.logger.Debugf("check-in stream failed to send expected state: %s", err)
-				}
-				return
-			}
-		}
-	}()
-
-	// at this point the client is connected, and it has sent it's first initial checkin
-	// the initial expected message must come before the sender goroutine will send any other
-	// expected messages. `CheckinExpected` method will also drop any expected messages that do not
-	// match the observed message to ensure that the expected that we receive is from the initial
-	// observed state.
 	c.initCheckinObservedMx.Lock()
-	c.initCheckinObserved = init
-	c.initCheckinExpectedCh = initExp
 	// clears the latest queued expected message
 	select {
 	case <-c.checkinExpected:
 	default:
 	}
+	c.initCheckinObserved = init
+	c.runtimeCheckinDone = make(chan struct{})
 	c.initCheckinObservedMx.Unlock()
+	defer func() {
+		close(c.runtimeCheckinDone)
+	}()
```

Comment:
I think you should assign this to a local variable as well that way the `defer func()` is always that channel. Otherwise it could happen (maybe not in current state of code, but in a future change) that `close` is not called on that channel if `c.runtimeCheckinDone` is assigned to a different channel.

```suggestion
	c.initCheckinObserved = init
	checkinDone := make(chan struct{})
	c.runtimeCheckinDone = checkinDone
	c.initCheckinObservedMx.Unlock()
	defer func() {
		close(checkinDone)
	}()
```

Reply:
Thanks, I see the point you’re making. I wasn’t initially sure what the best approach would be: leaving it as-is to let future test failures catch unintended changes, or guarding against them proactively. Since I couldn’t decisively lean one way, I went with your suggestion and assigned the channel to a local variable. I also passed it explicitly to the defer function to ensure it’s always the same channel being closed 🙂 respective commit b211ef16279f0835d21fbc655937f9451260a015